Don't profile ConstantScore queries (#23430)

hchokshi opened the pull request:
>A ConstantScore query will return the iterator of its inner query. However, when profiling, the constant score query is wrapped separately from its inner query, which distorts the times emitted by the profiler.
>
>This could be fixed by not wrapping the iterator if it is already a profiling implementation.

Added information regarding whether a search request has profiling available to the `QueryShardContext`, and used this to skip wrapping when profiling is active.

nik9000 commented on core/src/main/java/org/elasticsearch/index/query/ConstantScoreQueryBuilder.java:
```diff
@@ -138,7 +138,13 @@ public static ConstantScoreQueryBuilder fromXContent(QueryParseContext parseCont
+
+        if (context.isProfile()) {
+            // Skip wrapping if query is being profiled - issue 23430
+            return innerFilter;
```
I'm not super familiar with this code but my instinct is that this will change the scoring.

polyfractal commented:
I think this needs to be tackled during the profiling process, not during query construction for reasons mentioned by @nik9000 (it'll affect scoring if you just return the interior query, since that "unwraps" the ConstantScore and removes the constantly-scoring behavior).  

I think the appropriate place is likely in  [ContextIndexSearcher#createWeight()](https://github.com/kblincoe/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java#L113).  A weight is created for every component in the query, so this is where we wrap the query with a specialized `ProfileWeight`.  That profiling weight in turn creates a `ProfileScorer`, to capture scoring times too. 

So we can check to see if the query is a ConstantScore, and if it is, just return the regular weight without wrapping it in `ProfileWeight`.  That'll keep from "double-wrapping" the CS with two profiles.

I think :)  I haven't thought toooo deeply about this yet, but I believe that's the probably the correct route to take.

hchokshi commented:
I took the approach of checking if `weight` is already an instance of `ProfileWeight`, instead of checking if `query` is an instance of `ConstantScoreQuery` - in case there are any other situations now or in the future that would have a different type of query with an inner query.

joshhaller commented:
Just a small change but you could reduce the amount of comments eg you dont really need a line with "Issue 23430"

hchokshi commented:
@joshhaller How about now?

polyfractal commented on core/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java:
```diff
@@ -124,7 +124,13 @@ public Weight createWeight(Query query, boolean needsScores) throws IOException
             }
-            return new ProfileWeight(query, weight, profile);
+
+            if (weight instanceof ProfileWeight) {
```
I think this needs to happen before any profiling is done.  The profiler is a bit wonky in how it works, due to the way Lucene works: when query components are timed (via `startTime`) they are pushed onto a stack.  Lucene queries are hierarchical in nature, so calling `createWeight()` on one query will invoke `createWeight()` on any child queries.  The stack tracks the order and relationship between query components.

Once the method call returns, the time is recorded and the component popped off the stack.  After everything is done, the profiler uses the information gained during the stack push/pop to rebuild the tree structure of timnigs.

So, if a normal weight is returned instead of a `ProfileWeight`, the accounting at the end will likely get messed up because `n` times were recorded, but `n-1` weights were wrapped.

I believe if you move the `return weight` up before the `startTime()` it'll behave better.

hchokshi replied:
It's not possible to do this, because this would require `weight = super.createWeight(query, needsScores)` to run before `startTime()`, but this is the thing being timed.

polyfractal commented:
Depending on how far down the rabbit hole you want to go, and when this assignment is due, here's another note:

The current approach of skipping wrapping CS queries (or anything already profiled) works, but it does hide some information.  E.g. if you have this query `Bool( CS(Term), Term)`, the original code would generate a profile like:

```
- Bool(CS( Term), Term)
  - CS(Term)
    - Term
  - Term
```
Which is useful, because you'd like to see that the first `Term` is wrapped in a CS, and there is some minor timing to be done there (rewrites, etc).

Skipping the CS weight entirely turns the output into this:

```
- Bool( CS(Term), Term)
   - Term
   - Term
```
Which will still be mostly accurate because the bulk of the time is in the `Term`, not the CS, but it does omit some data.

I was talking to another Dev and the proper approach is probably to find a way to wrap the CS, but not time the DocIdSetIterator that is returned from the Scorer, since that's the part that delegates through from the internal query.  E.g. the ProfileScorer would not time the DocIdSetIterator for CS (or profiled) scorers.  Again, haven't thought too deeply how this would work, but wanted to mention it in case you were interested in digging a bit deeper.  :)

hchokshi commented:
@polyfractal In that case, would it work if I modified the `ProfileScorer` to do the following and reverted the change in `ContextIndexSearcher`.

```java
@Override
public DocIdSetIterator iterator() {
    // Unsure of this if condition - advise if it's incorrect
    if (scorer instanceof ProfileScorer || scorer instanceof ConstantScoreScorer) {
        return scorer.iterator();
    }

    final DocIdSetIterator in = scorer.iterator();
    return new DocIdSetIterator() { /* ... */ }
}
```

polyfractal commented:
Unsure, to be honest.   It seems reasonable, and probably what I would try as well, but not entirely certain.

We'd probably have to drag a lucene dev in here to confirm, this is starting to get to the edge of my knowledge about how the internals work.  @jpountz is on vacation this week, but would know more when he's back next week.
